pass-menu.1
.TH pass-menu 1 "Pass Menu" \fIpass-menu\fR(1) "Password Store Interface"

.SH NAME
pass-menu - A general frontend for password store

.SH SYNOPSIS
.B pass-menu
.RB [ OPTIONS ]
--
.B COMMAND
.RB [ ARGUMENTS ]

.SH DESCRIPTION
\fBpass-menu\fR is a command-line utility that provides a general
interface to password store that works well with any command that
accepts stdin. ie \fIfzf\fR(1), \fIdmenu\fR(1), \fIrofi\fR(1) and
even \fIgrep\fR(1).

\fBpass-menu\fR utilizes its own custom syntax for passfiles,
refer to \fBPASSFILE SYNTAX\fR section for detailed information.

.B NOTE:
This command should not be confused with \fBpassmenu\fR command that
ships by default with \fIpass\fR(1). The default \fBpassmenu\fR
command is very limited in functionality and is designed specifically
for \fIdmenu\fR(1).

.SH OPTIONS
.TP
\fB-t\fR,\fB --type\fR
Type the output, useful for filling text boxes in GUI applications.

.TP
\fB-c\fR,\fB --clip\fR
Copy the output to the clipboard and then restore the clipboard
after \fB45\fR (or \fBPASSWORD_STORE_CLIP_TIME\fR) seconds.

.TP
\fB-p\fR,\fB --print\fR
Print the output to stdout.

.TP
\fB-f\fR,\fB --filename=\fINAME\fR
Manually set the password store filename.

.TP
\fB-k\fR,\fB --key=\fINAME\fR
Manually set the password store key.

.TP
\fB-l\fR,\fB --logger=\fITYPE\fR
Set the logger type. Possible values are:

.TS
box nospaces tab(|);
Lb | L.
compact | Print POSIX errors and messages through stderr (default).
human   | Print human-readable errors and messages through stderr.
notify  | Print errors and messages through notifications.
.TE

.TP
\fB-F\fR,\fB --prompt-flag=\fIFLAG\fR
Flag passed to \fBCOMMAND\fR for prompting the user.

The values after short flag expand as described in the following table:

.TS
box nospaces tab(|);
L | L | L.
\fB-F\fIx\fR   | \fB-F\fR \fI-x\fR    | A single character is expanded to a short flag.
\fB-F\fIxxx\fR | \fB-F\fR \fI--xxx\fR | Multiple characters are expanded to a long flag.
.TE

.TP
\fB--file-prompt=\fIPROMPT\fR
Prompt message when choosing a password store filename.

Default value for different menu commands:

.TS
box nospaces tab(|);
Lb | L.
fzf      | "file> "
dmenu    | "file:"
default  | "file"
.TE

.TP
\fB--key-prompt=\fIPROMPT\fR
Prompt message when choosing a password store key.

Default value for different menu commands:

.TS
box nospaces tab(|);
Lb | L.
fzf      | "key> "
dmenu    | "key:"
default  | "key"
.TE

.TP
\fB--mode-prompt=\fIPROMPT\fR
Prompt message when choosing \fBpass-menu\fR mode.

Default value for different menu commands:

.TS
box nospaces tab(|);
Lb | L.
fzf      | "mode> "
dmenu    | "mode:"
default  | "mode"
.TE

.TP
\fB-h\fR,\fB --help\fR
Print the help message and exit.

.SH ENVIRONMENT VARIABLES
This section describes the environment variables that \fBpass-menu\fR
borrows from \fIpass\fR(1):

.TP
.B PASSWORD_STORE_DIR
Overrides the default password storage directory,
by default \fB~/.password-store/\fR.

.TP
.B PASSWORD_STORE_X_SELECTION
Overrides the selection passed to xclip, by default \fBclipboard\fR.
See \fIxclip\fR(1) for more info.

.TP
.B PASSWORD_STORE_CLIP_TIME
Specifies the number of seconds to wait before restoring the clipboard,
by default \fB45\fR seconds.

.SH EXAMPLES
.TP
.B FZF (Fuzzy Finder)

.EX
# Basic usage:
$ \fBpass-menu\fR -- \fBfzf\fR

# Copy a field from a password store file:
$ \fBpass-menu\fR --clip -- \fBfzf\fR

# Enable interactive input prompt:
$ \fBpass-menu\fR -Fprompt -- \fBfzf\fR
$ \fBpass-menu\fR --prompt-flag="--prompt" -- \fBfzf\fR
.EE

.TP
.B DMenu (Dynamic Menu)

.EX
# Basic usage:
$ \fBpass-menu\fR -- \fBdmenu\fR

# Type a field from a password store file:
$ \fBpass-menu\fR --type -- \fBdmenu\fR

# Enable interactive input prompt:
$ \fBpass-menu\fR -Fp -- \fBdmenu\fR
$ \fBpass-menu\fR --prompt-flag="-p" -- \fBdmenu\fR
.EE

.TP
.B Rofi

.EX
# Basic usage:
$ \fBpass-menu\fR -- \fBrofi\fR -dmenu

# Type a field from a password store file:
$ \fBpass-menu\fR --type -- \fBrofi\fR -dmenu

# Enable interactive input prompt:
$ \fBpass-menu\fR -Fp -- \fBrofi\fR -dmenu
$ \fBpass-menu\fR --prompt-flag="-p" -- \fBrofi\fR -dmenu
.EE

.TP
.B Non-Interactive

.EX
# Basic usage:
$ pass-menu --print --filename "GitHub" --key "Password"

# The above example also works with actions:
$ pass-menu --filename "GitHub" --key "((Autofill))"
.EE

.SH PASSFILE SYNTAX

This section describes the custom syntax used by the parser for
reading passfiles (password store files).

Here is an example password store file:

.TS
box;
L.
correct-horse-battery-staple
---
Username: hello-world
Email:    "${Username}@example.com"

otpauth://totp/hello@example.com?secret=MV4AU&issuer=Example

action(Autofill) :type Username :tab :type Password :clip OTP
.TE

A passfile supports four types of statements: the password line,
fields, actions, and otpauth URIs. Any line that doesn't contain one
of these statements is ignored by the parser.

The following sections go into more detail about each component of the syntax.

.TP
.B Password Line

The first line of the passfile gets treated as a field with \fBPassword\fR
as the key if it doesn't match a field, an otpauth URI, or an action.

.B EXAMPLES

The following table shows examples of password lines with their field
equivalent counterparts.

.TS
allbox nospaces tab(|);
Lb Lb
L L.
Password Line                | Field Counterpart
correct-horse-battery-staple | Password: correct-horse-battery-staple
Qnr9%5bZ.ZAurV7"?mqa         | Password: "Qnr9%5bZ.ZAurV7\\"?mqa"
.TE

.B EBNF DEFINITION

.EX
password-line = character+, "\\n";
.EE

.TP
.B Field

A key-value pair separated by a colon is regarded as a field.

The key must only contain the following characters \fB[-_a-zA-Z0-9<space><tab>]\fR
and cannot be \fBOTP\fR as it's reserved for otpauth URIs.

The value can either be raw text or a double quoted string.

Any leading or trailing whitespace is trimmed from raw text in field values.
If you want the whitespace, then the value should be quoted in a string.

.B EXAMPLES

.TS
allbox;
L.
Username: Example 123
Hello World: "hello ${Username}"
.TE

.B EBNF DEFINITION

.EX
field = whitespace*, field-key, whitespace*, ":", whitespace*, field-value, whitespace*, "\\n";

field-key = identifier - "OTP";

field-value = string | character+;
.EE

.TP
.B Action

Actions enable automation within \fBpass-menu\fR, performing tasks
such as autofilling forms, updating passwords, etc.

The action body consists of commands with an optional argument.
The argument can be either a single word or a string value.

.B COMMANDS

.TS
box nospaces tab(|);
Lb | Lb
L  | L.
Command                   | Description
_
\fB:tab\fR                | Press the tab key.
\fB:enter\fR              | Press the enter key.
\fB:type\fR \fI<REF>\fR   | Type the field or OTP that matches the given reference.
\fB:clip\fR \fI<REF>\fR   | Copy the field or OTP that matches the given reference to the clipboard.
\fB:run\fR \fI<REFS>\fR   | Execute comma separated list of the given action references.
\fB:log\fR \fI<STR>\fR    | Log the message with the given string.
\fB:sleep\fR \fI<DUR>\fR  | Delay execution for the given amount of time, accepts same arguments as \fIsleep\fR(1) command.
\fB:exec\fR \fI<CMD>\fR   | Execute the given bash command with $1 set to the current filename.
.TE

.B EXAMPLES

.TS
allbox;
L.
action(Autofill) :type Username :tab :type Password :clip OTP
action(PassGen) :log "Updating password" :exec "(pass-gen; echo; pass $1 | tail +2) >>(pass insert -m $1)"
action(Autofill and PassGen) :run Autofill,PassGen
.TE

.B EBNF DEFINITION

.EX
action = "action(", action-name, ")", whitespace+, (action-command, whitespace*)+, "\\n";

action-name = identifier;

action-command = action-command-name, [whitespace+, action-command-argument];

action-command-name = ":", ("tab" | "enter" | "type" | "clip" | "run" | "log" | "sleep" | "exec");

action-command-argument = string | (character - whitespace)+;
.EE

.TP
.B Otpauth URI

Any valid otpauth URI as described by Google Authenticator is allowed:

.UR https://github.com/google/google-authenticator/wiki/Key-Uri-Format
.UE

Multiple otpauth URIs with the same label are not allowed.

.B EXAMPLES

.TS
allbox;
L.
otpauth://hotp?counter=0&secret=JBSWY3DPEHPK3PXP
otpauth://totp/Example:alice@google.com?secret=JBSWY3DPEHPK3PXP&issuer=Example
.TE

.B EBNF DEFINITION

.EX
otpauth-uri = "otpauth://", otpauth-type, ["/", [otpauth-label]], "?", otpauth-parameters, "\\n";

otpauth-type = "totp" | "hotp";

otpauth-label = character+;

otpauth-parameters = otpauth-parameter, ("&", otpauth-parameter)*;

otpauth-parameter = letter+, "=", (character - "&")+;
.EE

.TP
.B String

Any double quoted value is treated as a string, strings can contain
escape characters and POSIX style variables with references to a field.

Allowed escape sequences are \fB\\\\\fR, \fB\\$\fR, and \fB\\"\fR.

.B EXAMPLES

.TS
allbox;
L.
"A normal string."
"A string with \\"escape\\" sequences."
"A string with reference to username: ${Username}."
"A string with reference to second username: ${Username[2]}"
.TE

.B EBNF DEFINITION

.EX
quote = "\\"";

string = quote, (character - quote | variable | escape)*, quote;

variable = "${", field-reference, "}";

escape = "\\", ("\\" | "$" | quote);
.EE

.TP
.B Identifier

An identifier is a name given to a field or an action. It must only
contain the following characters \fB[-_a-zA-Z0-9<space><tab>]\fR and
must not be entirely whitespace.

Multiple fields and actions can share the same identifier.

.B EBNF DEFINITION

.EX
identifier = (
    (letter | digit | "-" | "_")
    | (letter | digit | "-" | "_"),
      (letter | digit | "-" | "_" | " " | "\\t")*,
      (letter | digit | "-" | "_")
);
.EE

.TP
.B Reference

References are used to uniquely access data from a passfile.
There are 3 different types of references for accessing fields,
actions and otpauth URIs.

.B FIELD REFERENCE

A field reference begins with an identifier, excluding \fBOTP\fR, as
it is reserved for otpauth URIs. It can optionally be followed by a
natural number enclosed in square braces, indicating the index of the
field to use. If no index is provided, it defaults to the first field.

.TS
allbox nospaces tab(|);
Lb Lb
L L.
Example     | Description
Username    | Select the first field with the key "Username".
Username[1] | Same as above.
Username[2] | Select the second field with the key "Username".
.TE

.B ACTION REFERENCE

An action reference begins with an identifier enclosed in double round
braces. It can optionally be followed by its index enclosed in square
braces. If no index is provided, it defaults to the first action.

Here are some examples of action references:

.TS
allbox nospaces tab(|);
Lb Lb
L L.
Example         | Description
((Autofill))    | Select the first action with the name "Autofill".
((Autofill))[1] | Same as above.
((Autofill))[2] | Select the second action with the name "Autofill".
.TE

.B OTP REFERENCE

An OTP reference always begins with the \fBOTP\fR identifier. It can
optionally be followed by its label enclosed in round braces or its
index enclosed in square braces. If no label or index is provided, it
defaults to the first otpauth URI.

Here are some examples of otp references:

.TS
allbox nospaces tab(|);
Lb Lb
L L.
Example              | Description
OTP                  | Select the first otpauth URI.
OTP[1]               | Same as above.
OTP[2]               | Select the second otpauth URI.
OTP(user@github.com) | Select the otpauth URI with the label "user@github.com".
.TE

.B EBNF DEFINITION

.EX
index = "[", digit - "0", digit*, "]";

field-reference = field-key, [index];

action-reference = "((", action-name, "))", [index];

otp-reference = "OTP", [index | "(", otpauth-label, ")"];
.EE

.SH SEE ALSO
.MR pass 1 ,
.MR pass-gen 1 ,
.MR pass-genf 1

.SH AUTHOR
Udayvir Singh
.UR https://github.com/udayvir-singh
.UE

.SH LICENCE
MIT License

Copyright (c) Udayvir Singh

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


pass-menu.sh
#!/usr/bin/env bash

set +x
readonly VERSION="v0.0 git"

MODE=""
MENUCMD=""
TIMEOUT="45"
STOREDIR="${PASSWORD_STORE_DIR-$HOME/.password-store}"

if [ ${WAYLAND_DISPLAY} ]; then
	if type ydotool &>/dev/null; then
		DOTOOL="ydotool type --file /dev/stdin"
	else
		DOTOOL="wtype -"
	fi
	CLIPSET="wl-copy"
	CLIPGET="wl-paste"
else
	DOTOOL="xdotool type --clearmodifiers --file -"
	CLIPSET="xclip -selection clipboard"
	CLIPGET="${CLIPSET} -o"
fi

# ---------------------- #
#          INFO          #
# ---------------------- #
function show-help {
	echo \
"Usage: pass-menu [Option] -- COMMAND [ARGS]

Options:
  -p, --path              path to password store
  -c, --clip              copy output to clipboard
      --timeout           timeout for clearing clipboard [default: 45]
  -t, --type              type the output, useful in GUI applications 
  -e, --echo              print output to standard output
  -h, --help              display this help and exit
  -v, --version           output version information and exit

Examples:
	pass-menu --clip --timeout 15 -- fzf
	pass-menu --type              -- dmenu -i -l 15"

	exit 0
}

function show-version {
	echo "${VERSION}"
	exit 0
}

function error {
	local  MSG="${1}"; shift 1
	local ARGS="${@}"

	printf "pass-menu: ${MSG}\n" ${ARGS} >&2
	exit 1
}

# ---------------------- #
#          OPTS          #
# ---------------------- #
function set-opt-path {
	local OPTION="${1}"

	if [[ ! -d "${OPTION}" ]]; then
		error '--path "%s" is not a valid directory.' "${OPTION}" 
	fi

	STOREDIR="${OPTION}"
}

function set-opt-mode {
	local OPTION="${1}"

	if [[ "${MODE}" && "${MODE}" != "${OPTION}" ]]; then
		error 'conflicting option "--%s" with "--%s"' "${OPTION}" "${MODE}"
	fi

	MODE="${OPTION}"
}

function set-opt-timeout {
	local OPTION="${1}"

	if [[ ! "${OPTION}" =~ ^[0-9]+$ ]]; then
		error '"--timeout" must be an integer'
	elif [[ "${OPTION}" -lt 10 ]]; then
		error '"--timeout" must be greater than 10'
	fi
	
	TIMEOUT="${OPTION}"
}

function set-opt-cmd {
	local OPTION="${*}"

	MENUCMD="${OPTION}"
}

function opt-error {
	local OPTION="${1}"

	error 'invalid option "%s"' "${OPTION}"
}

while [ -n "${1}" ]; do
	case "${1}" in
		-p | --path)      set-opt-path "${2}"; shift 2 ;;
		-c | --clip)      set-opt-mode 'clip'; shift 1 ;;
		-t | --type)      set-opt-mode 'type'; shift 1 ;;
		-e | --echo)      set-opt-mode 'echo'; shift 1 ;;
		     --timeout)   set-opt-timeout "${2}" ; shift 2 ;;
		-h | --help)      show-help ;;
		-v | --version)   show-version ;;
		--)               shift; set-opt-cmd "${@}"; break ;;
		*)                opt-error "${1}" ;;
	esac
done

if [[ ! "${MENUCMD}" ]]; then
	error "missing required argument COMMAND"
fi

# ---------------------- #
#         UTILS          #
# ---------------------- #
function clip-copy {
	local VALUE="${1}"
	local ORIG="$(${CLIPGET})"
	local MSG="Copied to clipboard. Will clear in ${TIMEOUT} seconds."
	
	printf "${VALUE}" | ${CLIPSET}
	printf "%s\n" "${MSG}"

	if type notify-send &>/dev/null; then
		notify-send "${MSG}"
	fi

	{
		sleep "${TIMEOUT}" || exit 1
		# restore clipboard back to orginal if it hasn't changed.
		if [[ "$(${CLIPGET})" == "${VALUE}" ]]; then
			printf "${ORIG}" | ${CLIPSET}
		fi
	} &
}

function dotool-type {
	local VALUE="${1}"

	printf "${VALUE}" | ${DOTOOL}
}

# ---------------------- #
#          MAIN          #
# ---------------------- #
function get-pass-files {
	local LIST

	shopt -s nullglob globstar
	LIST=("$STOREDIR"/**/*.gpg)
	LIST=("${LIST[@]#"$STOREDIR"/}")
	LIST=("${LIST[@]%.gpg}")

	printf "%s\n" "${LIST[@]}" 
}

function get-pass-keys {
	local PASS_NAME="${1}"
	local PASS_FILE="$(pass "${PASS_NAME}")"
	
	if [[ ${#PASS_FILE} -lt 3 ]]; then
		error '"%s" is too short.' "${PASS_NAME}" 
	fi

	# Parse Action First
	awk '
	/action(.+)/ { 
		match($1, /action\((.+)\)/, a)
		printf "((%s))\n", a[1]
	}' <<< "${PASS_FILE}"

	# Parse Rest of Keys
	awk '
	BEGIN { 
		FS=": +"
		password="Yes" 
	} 

	NR == 1 && ! $2 { print "Password"; password=Null } 

	$2 { 
		sub("^ +", "", $1)
		if ( $1 == "Password") {
			if (password) print $1
		} else {
			print $1
		}
	}

	/^ *otpauth:/ { print "OTP" }' <<< "${PASS_FILE}"
}

function get-pass-value {
	local PASS_NAME="${1}"
	local PASS_KEY="${2}"

	case "${PASS_KEY}" in 
	OTP)      
		pass otp "${PASS_NAME}" 
	;;

	Password) 
		pass "${PASS_NAME}" | awk '
		BEGIN { 
			FS=": +"
			password="Yes" 
		} 

		NR == 1 && ! $2 { print $1; password=Null } 

		/Password/ && $2 { if (password) print $2 }'
	;;

	*) 
		pass "${PASS_NAME}" | awk -v key="${PASS_KEY}" '
		BEGIN { FS=": +" }

		$2 {
			if ($1 ~ key)
				for (i=2; i<=NF; i++) print $i
		}' 
	;;
	esac
}

function get-action {
	local PASS_NAME="${1}"
	local  ACT_NAME="${2}"

	pass "${PASS_NAME}" | awk -v action="action.+${ACT_NAME}" '{
		if ($1 ~ action) 
			for (i=2; i<=NF; i++) print $i
	}'
}

function execute-action {
	local PASS_NAME="${1}"
	shift

	while [ -n "${1}" ]; do
		case "${1}" in
		:clip)
			clip-copy "$(get-pass-value "${PASS_NAME}" "${2}")"
			shift 2
			;;
		:type)
			dotool-type "$(get-pass-value "${PASS_NAME}" "${2}")"
			shift 2
			;;
		:tab)
			dotool-type "	"
			shift 1
			;;
		:sleep)
			sleep "${2}"
			shift 2
			;;

		:exec | :notify)
			local ACT="${1}"
			local STR="${2:1}"
			shift 2
			# Parse String
			while [ ! "${STR:(-1)}" = '"' ]; do
				if [ -z "${1}" ]; then
					error 'unmatched {"} in %s.' "${PASS_NAME}"
				fi

				STR="${STR} ${1}"
				shift 1
			done

			STR="${STR::(-1)}"

			if [ "${ACT}" = ":exec" ]; then
				sh -c "${STR}"
			else
				notify-send "${STR}"
			fi
			;;

		:*) error "invalid action %s in %s" "${1}" "${PASS_NAME}" ;;
		*)  error "invalid param %s in %s" "${1}" "${PASS_NAME}" ;;
		esac
	done
}

function get-mode {
	if [[ "${MODE}" ]]; then
		printf "${MODE}"
	else
		local CANDIDATES="clip\ntype\necho"
		printf "${CANDIDATES}" | ${MENUCMD}
	fi
}

function call-menu {
	local PIPE="$(< /dev/stdin)"

	[ -z "${PIPE}" ] && exit 1

	printf "${PIPE}" | ${MENUCMD}
}

function main {
	local PASS_NAME PASS_KEY OUT

	PASS_NAME=$(get-pass-files | call-menu)
	[[ ! ${PASS_NAME} ]] && exit 1

	PASS_KEY=$(get-pass-keys "${PASS_NAME}" | call-menu)
	[[ ! ${PASS_KEY} ]] && exit 1

	if [ "${PASS_KEY:(-2)}" = "))" ]; then
		execute-action "${PASS_NAME}" $(get-action "${PASS_NAME}")
		return 0
	fi

	OUT=$(get-pass-value "${PASS_NAME}" "${PASS_KEY}")

	case "$(get-mode)" in
		clip) clip-copy "${OUT}" ;;
		type) dotool-type "${OUT}" ;;
		echo) printf "${OUT}" ;;
	esac
}

main
